Sources/NIO/Bootstrap.swift

```diff
@@ -325,6 +325,47 @@ private extension Channel {
     }
 }
 
+/// `ClientTransportBootstrap` is implemented by various underlying transport mechanisms. Typically,
+/// this will be the BSD Sockets API implemented by `ClientBootstrap`.
+public protocol ClientTransportBootstrap {
+
+    /// Create a `Bootstrap` on the `EventLoopGroup` `group`.
+    ///
+    /// - parameters:
+    ///     - group: The `EventLoopGroup` to use.
+    init(group: EventLoopGroup)
```

Comment:
In an ideal world we'd have a proper type here, but I am really struggling to come up with a way to do it. I'm going to think on it some more and maybe an idea will come to me.

Reply:
Agreed, please take your time 🙇‍♂️

I was wondering if we could use an associated type and force something, but I don't think that's correct. I'm hoping to figure out how to save folks from accidentally using a `MultiThreadedEventLoopGroup` with `NIOTSConnectionBootstrap. 🤔

Reply:
I think we should remove the `init` here and instead add a `public static func makeTCPClientBootstrap() -> NIOClientTransportBootstrap` to `EventLoopGroup`.

Sure, for API compatibility we need to implement a default implementation for this which would be

```
extension EventLoopGroup {
    public static func makeTCPClientBootstrap() -> NIOClientTransportBootstrap {
        struct CannotBootstrap: NIOClientTransportBootstrap {
            // fail everything with a good error
        }
        return CannotBootstrap()
    }
}
```

and in NIO 3 we can remove that. WDYT @Lukasa / @Yasumoto ?

Reply:
The reason for removing the `init` is because there's less room for error. The new way would basically be to choose your favourite `EventLoopGroup` and from there you could peel off a TCP client bootstrap.

Reply:
We would then just implement this for `SelectableEventLoop`, `MultiThreadedEventLoopGroup`, `EmbeddedEventLoop`, and `NIOTSEventLoop(Group)` which is 99.999% of the actual use-cases so nobody will ever see the error.

Reply:
I'm definitely open to doing that. An important note would be that we would _need_ to ensure that libraries don't ask for an event loop group, and instead ask for a bootstrap, when they don't care about which event loop is driving the bootstrap. The reason for that is that if you wanted to, for example, write a HTTP/2 client and server bootstrap that bootstraps stream channels, you wouldn't be hanging that off an event loop group.

It may be better to have a separate protocol to which all `EventLoopGroup`s conform that does this job. That would allow other types, such as `HTTP2StreamMultiplexer`, to expose conformances to that protocol as well. How does that seem?

Reply:
@Lukasa that sounds good but we would need to remove the `connect(...)` then, or would just have it fail for the H2 multiplexer use-case?

Reply:
That's a good question. In an ideal world `connect` would simply create a new stream, but the requirement that connect take a complex data type causes us some limitations. The biggest limitation there is that you need to put a host/port combo in the H2 multiplexer, which clearly doesn't work.

I don't know how to design our way around that either. In principle I guess we could use the host/port for authority information? But it raises some real questions about what all of this means that I don't think we can address today, so maybe we should drop that idea for now.

Reply:
Regardless, I think having a separate protocol is still the right choice, as it allows more flexibility for future evolution and implementations than tying the protocol directly to ELGs.

Reply:
@Lukasa ok, new idea

```
public struct NIOClientStreamBootstrap<ConnectType> {
     // all the methods

    public init<EL: EventLoop, ELG: EventLoopGroup>(group: ELG, eventLoopType: EL.Type = EL.self) { ... }

    public func connect(to: ConnectType)
}
```

plus

```
public protocol NIOClientStreamBootstrapper {
    associatedtype ConnectType

    // all the methods
}
```

```
extension ClientBootstrap: NIOClientStreamBootstrapper {
    typealias ConnectType = SocketAddress
}

extension HTTP2Multiplexer: NIOClientStreamBootstrapper {
    typealias ConnectType = H2StreamID
}
```

Reply:
that is more along the lines I was thinking before this PR came up but I never got to try it out to see if everything works... I'm happy to draft something more concrete up?

Reply:
the nice thing is that with `ConnectType`, we can also make `NIOPipeBootstrap` conform to it :)

Reply:
I think we'd need the connect type of `ClientBootstrap` to be an enum to allow for happy-eyeballing our way to victory, but that idea could work. The biggest downside is only that I suspect users will often constrain that associated type in a way that is somewhat troublesome. That was the advantage of requiring `host:port:` in the original implementation, it pushes you down the maximal feature route.

Reply:
@Lukasa what about

```
extension NIOClientStreamBootstrap where ConnectType == SocketAddress {
    public func connect(host: String, port: Int) {
        // happy eyeballs
    }
}
```

do you think this may work?

Reply:
It _does_, but that doesn't apply to NIOTransportServices.

Again, we can protocol our way out of this by having _another_ protocol that covers "can connect with host and port", or a type for "represents a host and port".

Reply:
(but besides that, I'd also be happy to make `ConnectType = SomeEnumThatCoversWhatWeHaveToday`

Reply:
At a certain point though I wonder if we're over-solving this. For now it may be sufficient to exclude everything but TCP clients and servers and to solve those cases today. We can try to generalise further later.

Reply:
@Lukasa agreed. We could just fly with basically what's there and name it accordingly leaving the more general name available for the next version?

Reply:
I would however remove the `init` and make users peel off the bootstrap from the ELG maybe?

Reply:
I'm fine with users peeling off from the ELG for now. Let's go down that road and worry about the best possible ergonomics later.

Reply:
sounds good to me! @Yasumoto are you okay with this and does this make sense?